ClosedXML/Excel/CalcEngine/XLCalcEngine.cs

```diff
@@ -23,6 +23,50 @@ public XLCalcEngine(IXLWorksheet ws) : this(ws.Workbook)
             _ws = ws;
         }
 
+        private IList<IXLRange> _cellRanges;
+
+        /// <summary>
+        /// Get a collection of cell ranges included into the expression. Order is not preserved.
+        /// </summary>
+        /// <param name="expression">Formula to parse.</param>
+        /// <returns>Collection of ranges included into the expression.</returns>
+        public IEnumerable<IXLRange> GetAffectingRanges(string expression)
+        {
+            _cellRanges = new List<IXLRange>();
+            Parse(expression);
+            var ranges = _cellRanges;
+            _cellRanges = null;
+            var visitedRanges = new HashSet<IXLRangeAddress>(new XLRangeAddressComparer(true));
+            foreach (var range in ranges)
+            {
+                if (!visitedRanges.Contains(range.RangeAddress))
+                {
+                    visitedRanges.Add(range.RangeAddress);
+                    yield return range;
+                }
+            }
+        }
+
+        public IEnumerable<IXLCell> GetAffectingCells(string expression)
+        {
+            if (String.IsNullOrWhiteSpace(expression) && String.IsNullOrEmpty(expression))
```

Comment:
This condition is really weird, isn't it?

Reply:
Sorry, I don't understand. What I meant was that you could just do this:
```c#
if (!String.IsNullOrWhiteSpace(expression) || !String.IsNullOrEmpty(expression))
{
    var ranges = GetAffectingRanges(expression);
    var visitedCells = new HashSet<IXLAddress>(new XLAddressComparer(true));
    var cells = ranges.SelectMany(range => range.Cells()).Distinct();
    foreach (var cell in cells)
    {
        if (!visitedCells.Contains(cell.Address))
        {
            visitedCells.Add(cell.Address);
            yield return cell;
        }
    }
}
```

Reply:
Oh, do you mean we could just do `if (!String.IsNullOrWhiteSpace(expression))` ?

Reply:
Exactly.

Reply:
Ok, so then that method's body could be just:
```c#
if (!String.IsNullOrWhiteSpace(expression))
{
    var ranges = GetAffectingRanges(expression);
    var visitedCells = new HashSet<IXLAddress>(new XLAddressComparer(true));
    var cells = ranges.SelectMany(range => range.Cells()).Distinct();
    foreach (var cell in cells)
    {
        if (!visitedCells.Contains(cell.Address))
        {
            visitedCells.Add(cell.Address);
            yield return cell;
        }
    }
}
```